### PR TITLE
🌱 Refactor tests to plain go in controllers

### LINKED ISSUE #4588: Refactor tests using plain go tests + envtest instead of ginkgo
As documented in https://cluster-api.sigs.k8s.io/developer/testing.html, we agreed to use plain go tests + envtest instead of ginkgo for all the integration tests (E2E test should continue to use ginkgo for the time being).

This issue is for tracking this effort.

/area testing
/kind cleanup
/help

### PR DESCRIPTION
**Pull request by mboersma**
**What this PR does / why we need it**:

Refactors the tests in the controllers package to use plain Go testing instead of ginkgo.

**Which issue(s) this PR fixes**:

Fixes #4588

**Notes**:

This is a big PR, although I think the changes fall into obvious patterns. But please let me know if I should break it up to make reviewing easier.

### CODE REVIEW

File: controllers/cluster_controller_test.go [hunk 1]
```diff
@@ -50,57 +48,59 @@ var _ = Describe("Cluster Reconciler", func() {
...
 		// Patch
-		Eventually(func() bool {
+		g.Eventually(func() bool {
 			ph, err := patch.NewHelper(cluster, testEnv)
-			Expect(err).ShouldNot(HaveOccurred())
+			g.Expect(err).NotTo(HaveOccurred())
 			cluster.Spec.InfrastructureRef = &v1.ObjectReference{Name: "test"}
 			cluster.Spec.ControlPlaneRef = &v1.ObjectReference{Name: "test-too"}
-			Expect(ph.Patch(ctx, cluster, patch.WithStatusObservedGeneration{})).ShouldNot(HaveOccurred())
+			g.Expect(ph.Patch(ctx, cluster, patch.WithStatusObservedGeneration{})).ShouldNot(HaveOccurred())
```
Comment (enxebre) [inline controllers/cluster_controller_test.go:98]: `To(Succeed())` reads better here

File: controllers/cluster_controller_test.go [hunk 2]
```diff
@@ -197,42 +202,44 @@ var _ = Describe("Cluster Reconciler", func() {
...
 
 		// Remove finalizers
-		Eventually(func() bool {
+		g.Eventually(func() bool {
 			ph, err := patch.NewHelper(cluster, testEnv)
-			Expect(err).ShouldNot(HaveOccurred())
+			g.Expect(err).NotTo(HaveOccurred())
 			cluster.SetFinalizers([]string{})
-			Expect(ph.Patch(ctx, cluster, patch.WithStatusObservedGeneration{})).ShouldNot(HaveOccurred())
+			g.Expect(ph.Patch(ctx, cluster, patch.WithStatusObservedGeneration{})).NotTo(HaveOccurred())
```
Comment (enxebre) [inline controllers/cluster_controller_test.go:235]: `To(Succeed())` reads better here

File: controllers/cluster_controller_test.go [hunk 3]
```diff
@@ -19,11 +19,7 @@ package controllers
 import (
 	"testing"
 
-	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
-	"sigs.k8s.io/cluster-api/util/conditions"
-
-	corev1 "k8s.io/api/core/v1"
 	v1 "k8s.io/api/core/v1"
```
Comment (enxebre) [inline controllers/cluster_controller_test.go:23]: should we instead keep the corev1 alias which is more meaningful and to avoid possible clashes?

Comment (mboersma) [reply controllers/cluster_controller_test.go:23]: Agreed, I'll use `corev1` instead.

File: controllers/cluster_controller_test.go (hunk 2 above)
Comment (mboersma) [reply controllers/cluster_controller_test.go:235]: It sure does--good catch! I replaced a few others like this with `To(Succeed())` but missed this one.

File: controllers/cluster_controller_test.go [hunk 4]
```diff
@@ -50,57 +48,59 @@ var _ = Describe("Cluster Reconciler", func() {
...
 	})
 
-	It("Should successfully patch a cluster object if the status diff is empty but the spec diff is not", func() {
+	t.Run("Should successfully patch a cluster object if the status diff is empty but the spec diff is not", func(t *testing.T) {
+		g := NewWithT(t)
+
 		// Setup
 		cluster := &clusterv1.Cluster{
 			ObjectMeta: metav1.ObjectMeta{
 				GenerateName: "test2-",
```
Comment (enxebre) [inline controllers/cluster_controller_test.go:73]: Orthogonal to this PR some tests here use `Namespace: "default"` whereas others use `Namespace:    v1.NamespaceDefault`, we should unify this and possibly use a test prefixed namespace for each. We might want to use a different PR for it. Thoughts?

Comment (vincepri) [reply controllers/cluster_controller_test.go:73]: Let's take it even a step further? We should generate a new namespace on the fly for each package, so tests cannot collide with each other

Comment (sbueringer) [reply controllers/cluster_controller_test.go:73]: @vincepri Do we need an additional issue to ensure we do this consistently in all our tests?

**Review by sbueringer — COMMENTED** [review]
lgtm except: https://github.com/kubernetes-sigs/cluster-api/pull/4615/files#r633605359

File: controllers/cluster_controller_test.go (hunk 4 above)
Comment (vincepri) [reply controllers/cluster_controller_test.go:73]: Yes that would be great

Comment (mboersma) [reply controllers/cluster_controller_test.go:73]: I created issue #4629 to track that.

**Review by vincepri — APPROVED** [review]
/approve

**Pull request merged**